models/classes/runner/time/QtiTimer.php

```diff
@@ -0,0 +1,340 @@
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA ;
+ */
+/**
+ * @author Jean-Sébastien Conan <jean-sebastien.conan@vesperiagroup.com>
+ */
+
+namespace oat\taoQtiTest\models\runner\time;
+
+use oat\taoTests\models\runner\time\InconsistentCriteriaException;
+use oat\taoTests\models\runner\time\InconsistentRangeException;
+use oat\taoTests\models\runner\time\InvalidDataException;
+use oat\taoTests\models\runner\time\InvalidStorageException;
+use oat\taoTests\models\runner\time\TimeException;
+use oat\taoTests\models\runner\time\TimeLine;
+use oat\taoTests\models\runner\time\TimePoint;
+use oat\taoTests\models\runner\time\TimeStorage;
+use oat\taoTests\models\runner\time\Timer;
+use qtism\runtime\tests\RouteItem;
+
+/**
+ * Class QtiTimer
+ * @package oat\taoQtiTest\models\runner\time
+ */
+class QtiTimer implements Timer
+{
+    /**
+     * The TimeLine used to compute the duration
+     * @var TimeLine
+     */
+    protected $timeLine;
+
+    /**
+     * The storage used to maintain the data
+     * @var TimeStorage
+     */
+    protected $storage;
+
+    /**
+     * QtiTimer constructor.
+     */
+    public function __construct()
+    {
+        $this->timeLine = new QtiTimeLine();
+    }
+
+    /**
+     * Adds a "server start" TimePoint at a particular timestamp for the provided ItemRef
+     * @param RouteItem $routeItem
+     * @param float $timestamp
+     * @return Timer
+     * @throws TimeException
+     */
+    public function start($routeItem, $timestamp)
+    {
+        // check the provided arguments
+        if (!($routeItem instanceof RouteItem)) {
+            throw new InvalidDataException('start() needs a valid routeItem instance!');
+        }
+        if (!is_numeric($timestamp) || $timestamp < 0) {
+            throw new InvalidDataException('start() needs a valid timestamp!');
+        }
+        
+        // extract the TimePoint identification from the provided item, and find existing range
+        $tags = $this->getItemTags($routeItem);
+        $range = $this->timeLine->find($tags, TimePoint::TARGET_SERVER);
+
+        // validate the data consistence
+        if (count($range) % 2) {
+            throw new InconsistentRangeException('The time range does not seem to be consistent, the range is not complete!');
```

Comment:
I'm not sure about that one, what if the browser crash. You'll have a start without end, or is it going to be a new occurence ?